ext/drmgr/drmgr.c

```diff
@@ -2497,3 +2510,105 @@ drmgr_disable_auto_predication(void *drcontext, instrlist_t *ilist)
     instrlist_set_auto_predicate(ilist, DR_PRED_NONE);
     return true;
 }
+
+/***************************************************************************
+ * EMULATION
+ */
+
+/* Reserve space for emulation note values */
+static void
+drmgr_emulation_init(void)
+{
+    note_base_emul = drmgr_reserve_note_range(DRMGR_NOTE_EMUL_COUNT);
+    ASSERT(note_base_emul != DRMGR_NOTE_NONE, "failed to reserve emulation note space");
+}
+
+/* Get note values based on emulation specific enums. Typically used by an
+ * instrumentation client running in conjunction with an emulation client.
+ */
+static ptr_int_t
+get_emul_note_val(int enote_val)
+{
+    return (ptr_int_t)(note_base_emul + enote_val);
+}
+
+/* Write emulation data to a label's data area. Typically set by an emulation
+ * client.
+ */
+static void
+set_emul_label_data(instr_t *label, int type, ptr_uint_t data)
+{
+    dr_instr_label_data_t *label_data = instr_get_label_data_area(label);
+    ASSERT(label_data != NULL, "failed to find label's data area");
+    label_data->data[type] = data;
+}
+
+/* Read emulation data from a label's data area. Typically read by an
+ * instrumentation client running in conjunction with an emulation client.
+ */
+static ptr_uint_t
+get_emul_label_data(instr_t *label, int type)
+{
+    dr_instr_label_data_t *label_data = instr_get_label_data_area(label);
+    ASSERT(label_data != NULL, "failed to find label's data area");
+    return label_data->data[type];
+}
+
+/* Set the start emulation label and emulated instruction data */
+DR_EXPORT
+void
+drmgr_create_emulation_start(void *drcontext, instrlist_t *ilist, instr_t *where,
+                             emulated_instr_t *einstr)
+{
+    instr_t *start_emul_label = INSTR_CREATE_label(drcontext);
+    instr_set_meta(start_emul_label);
+    instr_set_note(start_emul_label, (void *)get_emul_note_val(DRMGR_NOTE_EMUL_START));
+
+    set_emul_label_data(start_emul_label, DRMGR_EMUL_INSTR_PC, (ptr_uint_t)einstr->pc);
+    set_emul_label_data(start_emul_label, DRMGR_EMUL_INSTR, (ptr_uint_t)einstr->instr);
+    set_emul_label_data(start_emul_label, DRMGR_EMUL_ISA_VERSION,
+                        (ptr_uint_t)einstr->version);
+    einstr->fields = 3; /* fields <= data array size in dr_instr_label_data_t */
```

Comment:
The size field of `einstr` should be an input to this function, not written by this function.  It should fail if the size doesn't match what it expects (so maybe return a bool?), and if later we add new fields it would dispatch on the size to set the supported fields.

Reply:
Sorry Derek, I don't think I understand the logic for the size field, or I don't understand what your use case means. This is my understanding:
1 - We want to make sure that the size of emulated_instr_t does not exceed the label's data area.
2 - We (DynamoRIO developers) define the contents of emulated_instr_t to make sure it does not exceed the label's data area.
3 - To that end, we can check when label data is populated with emulated instruction data, in drmgr_insert_emulation_start(), that the size of emulated_instr_t does not exceed label's data area.

Is that not enough for our purposes? If we or any other developer needs to add elements to emulated_instr_t, the assert in drmgr_insert_emulation_start() should be enough to avoid errors. In other words, I don't understand how users of the function would be interested in sizing emulated_instr_t.

Reply:
No, the use case is purely to enable adding new fields in the future without breaking backward compatibility.  For any new struct introduced into the interface that is allocated by the user we typically want to use a size field to enable it to grow.  See how a size field is used in DR's existing exported data structures: dr_stats_t, dr_os_version_info_t, dr_memory_dump_spec_t, dr_syscall_result_info_t, drcovlib_options_t, drmodtrack_info_t, drmgr_priority_t, drreg_options_t, drsym_info_t, hashtable_config_t.  See also how it's used in the Windows API: e.g., https://docs.microsoft.com/en-us/windows/desktop/api/dbghelp/ns-dbghelp-_imagehlp_line.  It is a not-uncommon C interface mechanism.

Re: fitting in the label data area: I assumed that would be completely hidden from the user.  If we add more fields and they don't fit, at first I figured we'd use a pointer in the data area to some external memory we'd allocate on the heap -- but there's no hook to invoke us to clean it up, so we may have to instead create a 2nd label and point at that or sthg.  We'd still make that invisible to the user.

Reply:
For that freeing we'd have to document where the data goes away in case some observational client wanted to access it in the insru2instru phase.

We could try to add a free hook that DR would call: but it is not simple.  It would need to be per-label, so you'd think it'd have to take up one of the 4 slots.  Maybe that's the thing to do: add a new interface to set a free callback; store it in the 4th slot; set a flag saying whether the 4th slot holds that callback or not?

Reply:
I would expect debug build to fail with a leak assert with the current code -- does that not happen?  If not, how is the instr_t being freed in your own tests?  Maybe I'm missing something?

Reply:
> Hmm, on that note, how do we clean up the instr_t for the emulated instruction which is now removed from the instrlist and pointed at by the label data?

My (possibly naive?!) approach from the emulation client side is to create an `instr_t`, populate it with the emulated instruction encoding and put its pointer in the `emulated_instr_t` passed to `drmgr_insert_emulation_start()`, then remove from list and destroy, e.g.

Reply:
> Hmm, on that note, how do we clean up the instr_t for the emulated instruction which is now removed from the instrlist and pointed at by the label data?
. . .
> I would expect debug build to fail with a leak assert with the current code -- does that not happen? If not, how is the instr_t being freed in your own tests? Maybe I'm missing something?

No. My (possibly naive?!) approach from the emulation client side is not to point to the original `instr_t` but to create a new `instr_t`, populate it with the emulated instruction encoding and put its pointer in the `emulated_instr_t` passed to `drmgr_insert_emulation_start()`, then remove original `instr_t` from instrlist and destroy, e.g.

```C
. . .
  instr_t *einstr = instr_create(drcontext);                                                          
  instr_allocate_raw_bits(drcontext, einstr, <byte count>);                                                      
  instr_set_raw_bits(einstr, raw_instr_pc, <byte count>);                                                        
  instr_set_raw_bits_valid(einstr, true);
. . .
  emulated_instr_t emulated_instr = {raw_instr_pc, einstr, <ISA version>};                                        
  drmgr_insert_emulation_start(drcontext, bb, instr, &emulated_instr);
. . .
  instrlist_remove(bb, instr);                                                   
  instr_destroy(drcontext, instr);
. . .
```
This approach (i.e. completely removing the emulated instruction from instrlist and memory) avoids memory leaks and any issues with inadvertently executing it in a later phase, especially if the instruction is not implemented on the hardware, risking SIGILLs.

The idea behind this approach is based on the fact that DynamoRIO **may have no knowledge** of an emulated instruction, i.e. the emulation client decodes and builds the instruction in such a way that it can be executed by the client and observed by another.

Reply:
I'm asking about how the *new* instruction is cleaned up: `einstr` in your code snippet.  It's been allocated on the heap by `instr_create` but it is never freed.  Again, I would expect debug build to assert about a memory leak: does that not happen?

Reply:
It does not assert a memory leak on debug build with the code snippet above, but it does if the `instr_allocate_raw_bits()` call is removed. I removed it after seeing the following message (with the debug build):
 ```
Usage error: instr_set_raw_bits: bits already there, but different
```

Reply:
I've not tested on x86, only AArch64. Perhaps there is a difference? I'll try on an x86 box when I get a chance. In any case, it's definitely a leak and we need to clear up emulated instruction memory.

The question is which is the best approach: emulation client's responsibility via a list or DR's via a callback function pointer in the 4th slot? Which is the approach most consistent with the existing API?

The client side list doesn't use a slot which we may need later but requires more work by the user. The DR callback is less work for the user but occupies a slot which may result in expanding the label's data area later.

If we decide on the callback method, in which part of DR would we put the callback loop?

Reply:
The callback seems nicer to me.  If we add deeper-field capabilities to the label data area maybe we should include deep-copy support for cloning as well.

instr_free() would invoke the free callback for labels; instr_clone() would invoke the clone callback.  We could use a single callback and have it take in two instr_t* pointers: both are
non-NULL on a clone (old and new) while the 2nd is NULL on a free??  Too hacky?

For where to store the callback: annotations today use slots 0,1,2.  Could
take 3 and not break our own stuff, but for back compat safest to avoid
doing that.  Could add a flag, but we don't have much flag space.  Or we
could store it somewhere else like by adding a union with the raw bits
field (INSTR_RAW_BITS_ALLOCATED flag can distinguish which is which in that
new union).  So we'd add instr_set_label_callback(void (*cb)(instr_t *old,
instr_t *new)) and we'd assume a label has a callback if its `bytes` field is non-NULL
and INSTR_RAW_BITS_ALLOCATED is not set.

Reply:
Thanks for the pointers Derek. I'll have to pause this PR today, due to other commitments, but will do the freeing, the struct size and the test case as soon as I can.